presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java

```diff
@@ -121,54 +190,131 @@ public HivePartitionResult getPartitions(SemiTransactionalHiveMetastore metastor
-        return new HivePartitionResult(partitionColumns, partitionsIterable, compactEffectivePredicate, remainingTupleDomain, enforcedTupleDomain, hiveBucketHandle, bucketFilter);
+        return new HivePartitionResult(
+                partitionColumns,
+                table.getDataColumns(),
+                table.getParameters(),
+                partitions,
+                compactEffectivePredicate,
+                remainingTupleDomain,
+                enforcedTupleDomain,
+                hiveBucketHandle,
+                bucketFilter);
+    }
+
+    private Optional<HiveBucketHandle> getBucketHandle(
+            Table table,
+            ConnectorSession session,
+            TupleDomain<ColumnHandle> effectivePredicate)
+    {
+        // never ignore table bucketing for temporary tables as those are created such explicitly by the engine request
+        if (table.getTableType().equals(TEMPORARY_TABLE)) {
+            return getHiveBucketHandle(table);
+        }
+
+        Optional<HiveBucketHandle> hiveBucketHandle = getHiveBucketHandle(table);
+        if (!hiveBucketHandle.isPresent() || shouldIgnoreTableBucketing(session)) {
+            return Optional.empty();
+        }
+
+        if (queryUsesHiveBucketColumn(effectivePredicate)) {
+            return hiveBucketHandle;
+        }
+
+        int requiredTableBucketCount = getMinBucketCountToNotIgnoreTableBucketing(session);
+        if (hiveBucketHandle.get().getTableBucketCount() < requiredTableBucketCount) {
+            return Optional.empty();
+        }
+
+        return hiveBucketHandle;
+    }
+
+    private boolean queryUsesHiveBucketColumn(TupleDomain<ColumnHandle> effectivePredicate)
+    {
+        if (!effectivePredicate.getDomains().isPresent()) {
+            return false;
+        }
+        return effectivePredicate.getDomains().get().keySet().stream().anyMatch(key -> ((HiveColumnHandle) key).getName().equals(BUCKET_COLUMN_NAME));
+    }
+
+    private boolean queryAccessesTooManyBuckets(HiveBucketHandle handle, Optional<HiveBucketFilter> filter, List<HivePartition> partitions, ConnectorSession session)
+    {
+        int bucketsPerPartition = filter.map(hiveBucketFilter -> hiveBucketFilter.getBucketsToKeep().size())
+                .orElseGet(handle::getReadBucketCount);
+        return bucketsPerPartition * partitions.size() > getMaxBucketsForGroupedExecution(session);
+    }
+
+    private List<HivePartition> getPartitionsAsList(Iterator<HivePartition> partitionsIterator)
+    {
+        ImmutableList.Builder<HivePartition> partitionList = ImmutableList.builder();
+        int count = 0;
+        while (partitionsIterator.hasNext()) {
+            HivePartition partition = partitionsIterator.next();
+            if (count == maxPartitionsPerScan) {
+                throw new PrestoException(HIVE_EXCEEDED_PARTITION_LIMIT, format(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: maxPartitionsPerScan. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 301. Consider formatting this data with &#39;%d&#39; instead.